docs/Model Pipeline.md

```diff
@@ -0,0 +1,364 @@
+
+If a single model plugin existed, and all it did was map cloud instance types to microarchitectures and did that one job very well. Then other models may choose to rely on that one as a pre-requisite instead of implementing the functionality themselves:
+
+```yaml
+pipeline:
+  - instance-metadata
+  - ccf
+```
+
+> [!important] Ecosystem of plugins
+> The model plugins are the bulk of functionality for the impact engine. A small set of tightly focused model plugins can be combined in multiple ways to meet many use cases.
+
+### Upgradability
+
+Any time you upgrade one model plugin, every other plugin that depends on it can take advantage of those changes.
+
+Sticking to the above example of the `instance-metadata` plugin. A new version of the `instance-metadata` plugin can be released if a cloud provider updates its instance types or offerings. If you update your install, every plugin that depends on that data can take advantage of the update.
+### Consistency
+
+Again, sticking to the  `instance-metadata` plugin example. Breaking out functionality into lots of smaller plugins allows for more consistency. This is especially true regarding plugins that do lookups against data. 
+
+Rather than each plugin determining its own meta-data, one plugin can provide the metadata required for several subsequent plugins. From the data it exports to the observations, we can see what data every subsequent plugin is using.
+### Debuggability
+
+Since each model outputs a copy of its inputs, we can easily debug a calculation through its chain by dumping out the intermediate observations.
+### Simulation
+
+You can create models that **simulate** making changes to the system. For example, you could create a model called `change-instance-type`, which adjusts the data being passed through to **simulate** as if it was run on another cloud instance.
+
+Through a set of simulation plugins, you can investigate **what if scenarios**, see the impact of work before investing any time, like so
+
+```yaml
+component:
+  pipeline:
+    - change-instance-type 
+    - instance-metadata
+    - tdp
+    - teads  
+  config: 
+    change-instance-type:
+      to-instance: m5d.xlarge            
+  observations: 
+    - timestamp: 2023-07-06T00:00
+      duration: 5
+      cpu: 33
+      vendor: aws
+      instance_type: m5d.large
+```
+
+`change-instance-type` model would convert the above to this: 
+
+```yaml
+component:
+  pipeline:
+    - change-instance-type 
+    - instance-metadata
+    - tdp
+    - teads  
+  config: 
+    change-instance-type:
+      to-instance: m5d.xlarge            
+  observations: 
+    - timestamp: 2023-07-06T00:00
+      duration: 5
+      cpu: 17.5 # <-- updated
+      vendor: aws
+      instance_type: m5d.xlarge # <-- updated
+```
+
+`m5d.xlarge` is the same CPU but twice the size of `m5d.large`, so this plugin halves the utilization to mirror what might be on the new instance type.

```

Comment:
Here the utilization is an observation. It needs to come from the software only. What we could do is by changing the instance type using this simulation the impact model can output sim-m which is the embodied emissions change.

Reply:
There is still lots we need to think through around this, eventually I think there will form a dictionary of model terms.

Extract Model
Simulation Model
Adapter Model

I don't think there is any need to figure this out now but over time I think we'll evolve a language around this but it will be determined by the types of models that will be created.

RE: The confusion regarding **impacts** as a term for something that doesn't match our original defn of impacts. I agree but I'm not confident we have enough information yet to decide on an alternative approach? I suspect once we create more models and IMPLs then even more nuance might appear and an alternative language might surface.

Shall we aim to review this language around the decarb event and maybe our alpha release?

Reply:
Moving this discussion to an issue since I think it's not blocking the merging of this spec. https://github.com/Green-Software-Foundation/ief/issues/149